# Part 1: Forward Mode Automatic Differentiation

Forward mode AD can simply be implemented by defining a class to represent [dual numbers](https://en.wikipedia.org/wiki/Dual_number) which hold the value and its derivative. The following skeleton defines a dual number and implements multiplication. 

__Tasks:__

- Addition (`__add__`) is incomplete - can you finish it? 
- Can you also implement division (`__truediv__`), subtraction (`__sub__`) and power (`__pow__`)?

In [0]:
import math

class DualNumber:
    #'dunder method,let you emulate the behavior of built-in types'--https://dbader.org/blog/python-dunder-methods
    def __init__(self, value, dvalue):
        self.value = value
        self.dvalue = dvalue
    def __str__(self):

        return str(self.value) + " + " + str(self.dvalue) + "ε"
    def __mul__(self, other):
        return DualNumber(self.value * other.value,
            self.dvalue * other.value + other.dvalue * self.value)
    def __add__(self, other):
        return DualNumber(self.value + other.value,
            self.dvalue + other.dvalue)
    def __truediv__(self, other):
        return DualNumber(self.value / other.value,
            (self.dvalue * other.value - self.value * other.dvalue)/other.value**2) #https://en.wikipedia.org/wiki/Chain_rule#Quotient_rule
    def __sub__(self, other):
        return DualNumber(self.value - other.value,
            self.dvalue - other.dvalue)  
    #https://docs.python.org/3/library/operator.html
    #http://hyperphysics.phy-astr.gsu.edu/hbase/Math/derfunc.html
    def __pow__(self, other): 
        return DualNumber(self.value**other.value,
                        ((other.value*self.value**(other.value-1))*self.dvalue) + 
                        (self.value**other.value*math.log(self.value)*other.dvalue))         

In [0]:
# Tests

print(DualNumber(1,0) + DualNumber(1,0) / DualNumber(1,0) - DualNumber(1,0)**DualNumber(1,0))
# return a string representation of the object by The '__str__' method , either when someone codes in str(your_object), 
# or even when someone might do print(your_object). 


1.0 + 0.0ε


## Implementing math functions

We also need to implement some core math functions. Here's the sine function for a dual number:

In [0]:
# return an instance of DualNumber with the value of sin(x) and the dvalue of sin(x), and a x is also an instance of DualNumber 
def sin(x):
    return DualNumber(math.sin(x.value), math.cos(x.value)*x.dvalue)

__Task:__ can you implement the _cosine_ (`cos`), _tangent_ (`tan`), and _exponential_ (`exp`) functions in the code block below?

In [0]:
# TODO: implement additional math functions on dual numbers

def cos(x):
    return DualNumber(math.cos(x.value), -math.sin(x.value)*x.dvalue)

def tan(x):
    return DualNumber(math.tan(x.value), (1/math.cos(x.value)**2)*x.dvalue)

def exp(x):
    return DualNumber(math.exp(x.value), math.exp(x.value)*x.dvalue)

In [0]:
# Tests
assert cos(DualNumber(0,0)).value == 1
assert tan(DualNumber(0,0)).value == 0
assert exp(DualNumber(0,0)).value == 1


## Time to try it out

We're now in a position to try our implementation.

__Task:__ 

- Try running the following code to compute the value of the function $z=x\cdot y+sin(x)$ given $x=0.5$ and $y=4.2$, together with the derivative $\partial z/\partial x$ at that point. 

In [0]:
#to compute ∂z/∂x we just seed the algorithm with dx=∂x/∂x=1 and dy=∂y/∂x=0((since y and x are independent)) 
print(DualNumber(0.5,1)*DualNumber(4.2,0)+sin(DualNumber(0.5,1)))

2.579425538604203 + 5.077582561890373ε


__Task__: Differentiate the above function with respect to $x$ and write the symbolic derivatives in the following box. Verify the result computed above is correct by plugging-in the values into your symbolic gradient expression.

∂z/∂x or dz
<br>
=y*∂x/∂x+∂sin(x)/∂x or y*dx+dsin(x)
<br>
=y*1+cos(x)
<br>
=4.2+cos(0.5)
<br>
=5.07758256189037

__Task:__ Now use the code block below to compute the derivative $\partial z/\partial y$ of the above expression (at the same point $x=0.5, y=4.2$ as above) and store the derivative in the variable `dzdy` (just the derivative, not the Dual Number). Verify by hand that the result is correct.

In [0]:
dzdy=(DualNumber(0.5,0)*DualNumber(4.2,1)+sin(DualNumber(0.5,0))).dvalue

print('dz/dy:', dzdy)

dz/dy: 0.5


In [0]:
#Tests
assert dzdy
assert type(dzdy) == float


__Task:__ Finally, use the code block below to experiment and test the other math functions and methods you created.

In [0]:
dzdx=(DualNumber(0.5,1)*DualNumber(4.2,0)+tan(DualNumber(0.5,1))).dvalue

print('dz/dx:', dzdx)

dz/dx: 5.4984464104095245
